zanata-war/src/main/java/org/zanata/dao/ProjectMemberDAO.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2015, Red Hat, Inc. and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.zanata.dao;
+
+import org.hibernate.Query;
+import org.hibernate.Session;
+import org.jboss.seam.ScopeType;
+import org.jboss.seam.annotations.AutoCreate;
+import org.jboss.seam.annotations.Name;
+import org.jboss.seam.annotations.Scope;
+import org.zanata.model.HPerson;
+import org.zanata.model.HProject;
+import org.zanata.model.HProjectMember;
+import org.zanata.model.ProjectRole;
+
+import java.util.HashSet;
+import java.util.Set;
+
+
+/**
+ * Provides methods to access data related to membership in a project.
+ */
+@Name("projectMemberDAO")
+@AutoCreate
+@Scope(ScopeType.STATELESS)
+public class ProjectMemberDAO
+        extends AbstractDAOImpl<HProjectMember, HProjectMember.HProjectMemberPK> {
+
+    public ProjectMemberDAO() {
+        super(HProjectMember.class);
+    }
+
+    public ProjectMemberDAO(Session session) {
+        super(HProjectMember.class, session);
+    }
+
+    /**
+     * Retrieve all of a person's roles in a project.
+     */
+    public Set<ProjectRole> getRolesInProject(HPerson person, HProject project) {
+        Query query = getSession().createQuery(
+                "from HProjectMember as m where m.person = :person " +
+                        "and m.project = :project")
+                .setParameter("person", person)
+                .setParameter("project", project)
+                .setComment("ProjectMemberDAO.getRolesInProject");
```

Comment:
I think hibernate query cache will handle the cache update. We don't need
to programatically clear or invalidate cache entries.

On Thu, Aug 27, 2015 at 4:25 PM, David Mason <notifications@github.com>
wrote:

> In zanata-war/src/main/java/org/zanata/dao/ProjectMemberDAO.java
> <https://github.com/zanata/zanata-server/pull/948#discussion_r38065617>:
>
> > +    }
> > +
> > +    public ProjectMemberDAO(Session session) {
> > +        super(HProjectMember.class, session);
> > +    }
> > +
> > +    /**
> > +     * Retrieve all of a person's roles in a project.
> > +     */
> > +    public Set<ProjectRole> getRolesInProject(HPerson person, HProject project) {
> > +        Query query = getSession().createQuery(
> > +                "from HProjectMember as m where m.person = :person " +
> > +                        "and m.project = :project")
> > +                .setParameter("person", person)
> > +                .setParameter("project", project)
> > +                .setComment("ProjectMemberDAO.getRolesInProject");
>
> @huangp <https://github.com/huangp> if I make it cacheable, I will also
> want to clear the cache whenever there is a change to someone's locale
> membership. Is there a good way to do that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/948/files#r38065617>.
>



-- 
Patrick Huang
Senior Software Engineer
Engineering - Internationalisation
Red Hat, Asia-Pacific Pty Ltd
Level 1, 193 North Quay
Brisbane 4000
Office: +61 7 3514 8278
Fax: +61 7 3514 8199
IRC: pahuang
github: github.com/huangp
Website: www.redhat.com